modules/transport-netty4/src/main/java/org/opensearch/http/reactor/ReactorNetty4HttpServerTransport.java

```diff
@@ -0,0 +1,325 @@
+        NettyAllocator.logAllocatorDescriptionIfNeeded();
+        this.readTimeoutMillis = Math.toIntExact(SETTING_HTTP_READ_TIMEOUT.get(settings).getMillis());
+    }
+
+    @Override
+    protected HttpServerChannel bind(InetSocketAddress socketAddress) throws Exception {
+        HttpServer server = HttpServer.create()
+            .bindAddress(() -> socketAddress)
+            .wiretap(true)
+            .compress(true)
+            .protocol(HttpProtocol.HTTP11, HttpProtocol.H2C)
+            .handle((req, res) -> incomingRequest(req, res));
+
+        disposableServer = server.bindNow();
+        return new ReactorNetty4HttpServerChannel(disposableServer.channel());
+    }
+
+    @Override
+    protected void stopInternal() {
+        if (disposableServer != null) {
+            disposableServer.disposeNow();
+        }
+    }
+
+    @Override
+    public void onException(HttpChannel channel, Exception cause) {
+        if (cause instanceof ReadTimeoutException) {
+            super.onException(channel, new HttpReadTimeoutException(readTimeoutMillis, cause));
+        } else {
+            super.onException(channel, cause);
+        }
+    }
+
+    @Override
+    protected void doStart() {
+        boolean success = false;
+        try {
+            bindServer();
+            success = true;
+        } finally {
+            if (success == false) {
+                doStop(); // otherwise we leak threads since we never moved to started
+            }
+        }
+    }
+
+    protected Publisher<Void> incomingRequest(HttpServerRequest request, HttpServerResponse response) {
+        final Method method = HttpConversionUtil.convertMethod(request.method());
+        if (dispatcher.dispatchAsStream(request.uri(), request.fullPath(), method, request.params())) {
+            final StreamingRequestConsumer<HttpContent> consumer = new StreamingRequestConsumer<>(request, response);
+
+            request.receiveContent().defaultIfEmpty(DefaultLastHttpContent.EMPTY_LAST_CONTENT).subscribe(consumer);
+
+            return response.sendObject(consumer);
+        } else {
+            final NonStreamingRequestConsumer<HttpContent> consumer = new NonStreamingRequestConsumer<>(request, response);
+
+            request.receiveContent().defaultIfEmpty(DefaultLastHttpContent.EMPTY_LAST_CONTENT).subscribe(consumer);
+
+            return Mono.from(consumer).flatMap(hc -> {
+                // TODO: Copy other response details.
+                final FullHttpResponse r = (FullHttpResponse) hc;
+                response.status(r.status());
+                r.headers().forEach(h -> response.addHeader(h.getKey(), h.getValue()));
+                return Mono.from(response.chunkedTransfer(false).sendHeaders().sendObject(r.content()));
+            });
+        }
+    }
+
+    private class StreamingRequestConsumer<T extends HttpContent> implements Consumer<T>, Publisher<HttpContent> {
```

Comment:
The security plugin works by intercepting requests through the transport layer, bulk requests are skipped since there individual index requests are fanned out [1].  I suspect this will work correctly for streaming requests assuming the requests are still fanned out via transport actions - if they aren't a new hook will be needed.  If a new hook is needed, it could be added by using the IdentityService [2] to get the subject and then perform the permissions check.

After this is merged into main, but before its backported, I would highly recommend adding a new test within the security plugin to exercise the Streaming API an ensure permissions conventions are not side-stepped.

- [1] https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L259-L262
- [2] https://github.com/opensearch-project/OpenSearch/pull/7246/files#diff-aac303dafc3ff8f2c94c4913c62dc5f57e3ee1c3dd2633188a956da68697fa52